### PR TITLE
feat(xrpc): register variable rpc_time & redis_cmd_line

### PR DESCRIPTION
**Pull request by tzssangglass**
### Description

Fixes # (issue)

### Checklist

- [x] I have explained the need for this PR and the problem it solves
- [ ] I have explained the changes or the new features added to this PR
- [x] I have added tests corresponding to this change
- [ ] I have updated the documentation to reflect this change
- [ ] I have verified that this change is backward compatible (If not, please discuss on the [APISIX mailing list](https://github.com/apache/apisix/tree/master#community) first)

### CODE REVIEW

File: docs/en/latest/apisix-variable.md [hunk 1]
```diff
@@ -39,5 +39,7 @@ List in alphabetical order:
 | route_name       | core    | name of `route`        |   |
 | service_id       | core    | id of `service`        |   |
 | service_name     | core    | name of `service`      |   |
+| rpc_time         | xRPC    | time spent at the rpc request level |   |
+| redis_cmd_line   | redis   | the type of Redis command |   |
```
Comment (spacewander) [inline docs/en/latest/apisix-variable.md:43]:
```suggestion
| redis_cmd_line   | Redis   | the content of Redis command |   |
```

File: docs/en/latest/xrpc.md [hunk 2]
```diff
@@ -131,6 +131,52 @@ One specifies the `superior_id`, whose corresponding value is the ID of another
...
+                }
+            }
+        }
+    }
+}
+```
+
+This configuration means that when the `rpc_time` is greater than 10 ms, xPRC reports the request log to the log server via the `syslog` plugin. `conf` is the configuration of the logging server required by the `syslog` plugin.
+
+Unlike standard TCP proxies, which only execute a logger when the connection is broken, xRPC's executed logger at the end of each 'request'.
```
Comment (spacewander) [inline docs/en/latest/xrpc.md:169]:
```suggestion
Unlike standard TCP proxies, which only execute a logger when the connection is closed, xRPC's executed logger at the end of each 'request'.
```

File: docs/en/latest/xrpc.md [hunk 3]
```diff
@@ -131,6 +131,52 @@ One specifies the `superior_id`, whose corresponding value is the ID of another
...
+
+The protocol itself defines the granularity of the specific request, and the xRPC extension code implements the request's granularity.
+
+For example, in the Redis protocol, the execution of a command is considered a request.
+
+### Exposes variables
```
Comment (spacewander) [inline docs/en/latest/xrpc.md:175]: We can omit this section as it is written in docs/en/latest/apisix-variable.md

File: apisix/stream/xrpc/runner.lua [hunk 4]
```diff
@@ -27,6 +27,19 @@ local pcall = pcall
...
+core.ctx.register_var("rpc_time", function(ctx)
+    local session = ctx.xrpc_session
+    local curr_ctx_id = session._currr_ctx_id
+    local curr_ctx = session._ctxs[curr_ctx_id]
+
+    if not curr_ctx then
+        core.log.warn("can't find current context by id: ", curr_ctx_id)
+        return nil
+    end
+    return curr_ctx._rpc_end_time * 1000 - curr_ctx._rpc_start_time * 1000
```
Comment (spacewander) [inline apisix/stream/xrpc/runner.lua:40]: Let's use second as the unit, like the `request_time`

File: apisix/stream/xrpc/runner.lua [hunk 5]
```diff
@@ -27,6 +27,19 @@ local pcall = pcall
 local ipairs = ipairs
 local tostring = tostring
 
+
+core.ctx.register_var("rpc_time", function(ctx)
+    local session = ctx.xrpc_session
+    local curr_ctx_id = session._currr_ctx_id
+    local curr_ctx = session._ctxs[curr_ctx_id]
```
Comment (spacewander) [inline apisix/stream/xrpc/runner.lua:34]: I think we can implement a special var meta for the xRPC ctx?
A new version of https://github.com/apache/apisix/blob/4690feb421779f5b79e8dd990dc00f4d3f1052d0/apisix/core/ctx.lua#L216.

Therefore we can avoid looking up current xRPC ctx in every vars.

BTW, the `release_vars` is not called for xRPC ctx.

Comment (spacewander) [reply apisix/stream/xrpc/runner.lua:34]: Or we can pass ctx in https://github.com/apache/apisix/blob/4690feb421779f5b79e8dd990dc00f4d3f1052d0/apisix/core/ctx.lua#L343 as var._ctx

File: docs/en/latest/apisix-variable.md (hunk 1 above)
Comment (tzssangglass) [reply docs/en/latest/apisix-variable.md]: update

File: docs/en/latest/xrpc.md (hunk 2 above)
Comment (tzssangglass) [reply docs/en/latest/xrpc.md]: update

File: docs/en/latest/xrpc.md (hunk 3 above)
Comment (tzssangglass) [reply docs/en/latest/xrpc.md]: remove this

File: apisix/stream/xrpc/runner.lua (hunk 4 above)
Comment (tzssangglass) [reply apisix/stream/xrpc/runner.lua]: update

File: apisix/stream/xrpc/runner.lua (hunk 5 above)
Comment (tzssangglass) [reply apisix/stream/xrpc/runner.lua]: due to:https://github.com/apache/apisix/blob/4690feb421779f5b79e8dd990dc00f4d3f1052d0/apisix/core/ctx.lua#L345

in `core.ctx.register_var`, the `ctx.var._ctx` would point to `ngx.ctx.api_ctx`, so I choose  to pass th xRPC ctx in `session._curr_ctx`

File: apisix/stream/xrpc/runner.lua [hunk 6]
```diff
@@ -27,6 +27,19 @@ local pcall = pcall
...
+
+core.ctx.register_var("rpc_time", function(ctx)
+    local curr_ctx = ctx.xrpc_session and ctx.xrpc_session._curr_ctx
+
+    if not curr_ctx then
+        core.log.warn("can't find current context by id: ", curr_ctx_id)
+        return nil
+    end
+    --use second as the unit, like the request_time
+    local time = (curr_ctx._rpc_end_time * 1000 - curr_ctx._rpc_start_time * 1000) / 1000
```
Comment (spacewander) [inline apisix/stream/xrpc/runner.lua:39]: Why do we multiply 1000 first and divide by 1000 later?

File: t/xrpc/redis.t [hunk 7]
```diff
@@ -565,3 +577,126 @@ passed
 --- response_body
 ok
 --- stream_conf_enable
+
+
+
+=== TEST 12: set custom log format
```
Comment (spacewander) [inline t/xrpc/redis.t:583]: Maybe we can split the redis variable into another PR and the test can be put into `redis2.t`. Otherwise, there will be merge conflict.

Comment (tzssangglass) [reply t/xrpc/redis.t:583]: split this PR, so close.

File: apisix/stream/xrpc/runner.lua (hunk 5 above)
Comment (spacewander) [reply apisix/stream/xrpc/runner.lua]: @tzssangglass 
The `register_var` doesn't rewrite the `ctx.var`. `ctx.var._ctx` will point to the ctx passed via `set_vars_meta(ctx)`.
